File: functions/Hinich_my.py
"""
Тест Хинича на заданной системе скользящих окон
"""
import numpy as np
from .GLSTAT import GLSTAT
from utils import setup_figure, finalize_figure, iter_windows
from config import config


def Hinich_my(Y, start_offset, window_shift, window_width, num_windows, fig=None):
    """
    Выполнение теста Хинича на системе скользящих окон.
    """
    Y = np.array(Y).flatten()
    Y1 = np.diff(Y, n=2)
    s = config.plots.style
    c = config.constants

    P_gauss = []
    del_R = []

    for window in iter_windows(Y1, start_offset, window_shift, window_width, num_windows):
        sg, sl = GLSTAT(window, c.hinich_threshold, c.hinich_bins)
        P_gauss.append(sg[2])
        del_R.append(abs(sl[0] - sl[2]))

    actual_windows = len(P_gauss)

    fig, [ax1, ax2], standalone = setup_figure(fig, 2, 1, figsize=(12, 8))

    ax1.bar(range(1, actual_windows + 1), P_gauss, linewidth=2)
    ax1.grid(True)
    ax1.set_title('Тест Хинича',
                  fontsize=s.title_fontsize + 1, fontweight=s.title_fontweight)
    ax1.set_ylabel('P_gauss',
                   fontsize=s.label_fontsize, fontweight=s.label_fontweight)
    ax1.set_xlim([0, actual_windows + 1])
    ax1.set_ylim([0, 1.1])

    ax2.bar(range(1, actual_windows + 1), del_R, linewidth=2)
    ax2.grid(True)
    ax2.set_xlim([0, actual_windows + 1])
    max_val = max(del_R) if del_R else 0
    ax2.set_ylim([0, max_val * 1.1 if max_val > 0 else 5])
    ax2.set_ylabel('del_R',
                   fontsize=s.label_fontsize, fontweight=s.label_fontweight)
    ax2.set_xlabel('Номер скользящего окна',
                   fontsize=s.label_fontsize, fontweight=s.label_fontweight)

    finalize_figure(fig, standalone)

File: matlab/Hinich_my.m
function Hinich_my(Y,K,d,L,R);
%Тест Хинича на заданной системе скользящих окон 
Y1=diff(Y,2);  
for k=1:R;
      X=Y1(K+d*(k-1)+1:K+d*(k-1)+L,:);     
      [sg, sl]=GLSTAT(X,0.51,128);
      
      P_gauss(k)=sg(3);   del_R(k)=abs(sl(1)-sl(3));  
  end;
  hF3=figure; 
  set(hF3,'Units','Normalized','Pos',[0.2 0.175 0.78 0.55],...
    'Color',[0.8 0.8 0.8]);
  subplot(2,1,1);  bar(P_gauss,'LineWidth',2); grid;
title('Тест Хинича','FontSize', 15,'FontWeight','bold');
ylabel('P_gauss','FontSize', 14, 'FontWeight', 'bold');
axis([0 R 0 1.1]);
  subplot(2,1,2);  bar(del_R,'LineWidth',2); grid;
axis([0 R 0 5]);
ylabel('del_R','FontSize', 14, 'FontWeight', 'bold');
xlabel('Номер скользящего окна','FontSize',14,'FontWeight','bold');
  